### ISSUE TITLE
Can docker environments be used when training requires a gpu?

### ISSUE THREAD

**Issue by greaber** [mlflow/mlflow#1093]
To use the gpu in docker, we need to pass the `--runtime=nvidia` option to `docker run`.  But apparently mlflow doesn't provide any way to customize the `docker run` command.  So does that mean if you want to use a gpu you are limited to conda environments (or hacks like changing the system-wide default docker runtime)?

**Comment by samgd** [comment]
Also hitting this issue.

**Comment by ginochen** [comment]
Same here, have you resolved this issue?

**Comment by adeandrade** [comment]
My solution so far has been to run `mlflow run` inside the docker container (with `--no-conda` and no docker container specified in the project definition).

Effectively bypassing the feature, I guess.

**Comment by samgd** [comment]
Still a blocker unless an update has solved it.

**Comment by nlml** [comment]
I found a workaround following [this stackoverflow answer](https://stackoverflow.com/a/47495905/6167850) - just set your default runtime by adding first-level key to `/etc/docker/daemon.json` `"default-runtime": "nvidia"`, then `sudo service docker restart`

**Comment by izanagi1995** [comment]
We are also interested. We don't want to set the default runtime to nvidia also..

**Comment by pogil** [comment]
This would be a good first issue. We could extend the MLproject spec to support a new `runtime` field in the `docker_env` configuration block. E.g.
```
docker_env:
  image: mlflow-docker-example-environment
  runtime: nvidia
```
then update `_get_docker_command` in `mlflow/projects/_init.py` and its [invocation](https://github.com/mlflow/mlflow/blob/889e46269256c3bef5c9744312fd860c0136521a/mlflow/projects/__init__.py#L154) to process the `runtime` parameter and add it to the command arguments. If someone picks this up, it would be great to also update the documentation.

**Comment by ksanjeevan** [comment]
+1

It would be great in general if we could just pass arbitrary `docker run` arguments (like I would pass `-gpus all`). Having MLflow handle training without GPU support just doesn't make sense without it...

**Comment by Zethson** [comment]
Yeah, this is mandatory for me. Hoping to see some progress here.

**Comment by pogil** [comment]
@ksanjeevan good points - perhaps instead of a `runtime` parameter, we should instead allow `additional-arguments` or something similar. We would love a community contribution on this issue!

**Comment by Zethson** [comment]
@pogil Maybe this issue should be merged with https://github.com/mlflow/mlflow/issues/2607 ?

**Comment by ksanjeevan** [comment]
@pogil would something like https://github.com/mlflow/mlflow/pull/2608 seem reasonable? I have it as a draft as I didn't want to start writing tests/doc before getting more feedback on the approach.

**Comment by ksanjeevan** [comment]
FYI - addressed in https://github.com/mlflow/mlflow/pull/2608

**Comment by tomasatdatabricks** [comment]
As @ksanjeevan mentioned the issue has been addressed by #2608.
